test/format/config.c

```diff
@@ -275,36 +276,29 @@ config_cache(void)
 			    g.c_cache_minimum, g.c_cache);
 		return;
 	}
-
-	/* Check if a minimum cache size has been specified. */
 	if (g.c_cache_minimum != 0 && g.c_cache < g.c_cache_minimum)
 		g.c_cache = g.c_cache_minimum;
 
-	/* Ensure there is at least 1MB of cache per thread. */
-	if (g.c_cache < g.c_threads)
-		g.c_cache = g.c_threads;
-
 	/*
 	 * Maximum internal/leaf page size sanity.
 	 *
 	 * Ensure we can service at least one operation per-thread concurrently
 	 * without filling the cache with pinned pages, that is, every thread
-	 * consuming an internal page and a leaf page. Page-size configurations
-	 * control on-disk sizes and in-memory pages are often larger than their
-	 * disk counterparts, so it's hard to translate from one to the other.
-	 * Use a size-adjustment multiplier as an estimate.
+	 * consuming an internal page and a leaf page (or a pair of leaf pages
+	 * for cursor movements).
 	 *
 	 * Assuming all of those pages are dirty, don't let the maximum dirty
 	 * bytes exceed 40% of the cache (the default eviction trigger is 20%).
+	 *
+	 * Maximum memory pages are in units of MB.
+	 *
+	 * This code is what dramatically increases the cache size when there
+	 * are lots of threads, it grows the cache to N megabytes per thread.
 	 */
-#define	SIZE_ADJUSTMENT	3
-	for (;;) {
-		max_dirty_bytes = ((g.c_cache * WT_MEGABYTE) / 10) * 4;
-		if (SIZE_ADJUSTMENT * g.c_threads *
-		    (g.intl_page_max + g.leaf_page_max) <= max_dirty_bytes)
-			break;
-		++g.c_cache;
-	}
+	v = 2 * g.c_threads * MEGABYTE(g.c_memory_page_max);
+	g.c_cache = WT_MAX(g.c_cache, (v + (WT_MEGABYTE - 1)) / WT_MEGABYTE);
+	v = ((g.c_cache * WT_MEGABYTE) / 10) * 4;
+	g.c_cache = WT_MAX(g.c_cache, (v + (WT_MEGABYTE - 1)) / WT_MEGABYTE);
```

Comment:
I don't understand the last two lines: don't they boil down to `cache = WT_MAX(cache, 0.4 * cache)` -- i.e., it's always going to be smaller than the first calculation, so won't have any effect?

Reply:
Yes, I did that calculation incorrectly, I've pushed a change in 30cb00f that I think gets it right.

We talked about this briefly: I recall @agorrod made that change, configuring the total number of dirty pages by all the threads to never exceed 40% of the cache. Maybe he remembers his thinking on this.

Regardless, this change as written, is going to significantly increase the size of pretty much every cache we run with. I think I'm OK with that, given the relative maturity of WiredTiger and the kinds of debugging we're doing (vs. the kind of debugging we were doing a few years ago).

Reply:
I still don't really understand the adjustment to `v` - it's doing a calculation based on threads and maximum page sizes, then multiplying that by 2.5. Isn't that the same as basing the calculation off 5 pages? Is what we want here to actually use 3 - 2 leaf pages and an internal page? That assumes a tree depth of 1 (plus root level), but I think it's reasonable.

Oh - I see. It's adjusting so that those pages can be dirty. Could we instead use 3 here, and set a higher `eviction_dirty_trigger` like we already do for LSM trees? Do we expect test/format to meaningfully want or test different clean/dirty targets?

Or use a higher constant multiplier? The dirty trigger is a hard limit, so using 40% seems a bit weird (I doubt it was a hard limit when I made the change originally, or maybe I just didn't think hard about it).